TowerForge/TowerForge/GameModule/Systems/BaseSystems/ContactSystem.swift

```diff
@@ -9,59 +9,80 @@ import Foundation
 
 class ContactSystem: TFSystem {
     var isActive = true
-    weak var entityManager: EntityManager?
-    weak var eventManager: EventManager?
-
-    private var contacts: Set<TFContact> = []
+    unowned var entityManager: EntityManager
+    unowned var eventManager: EventManager
+    private var contacts: Set<TFContact>
 
     init(entityManager: EntityManager, eventManager: EventManager) {
         self.entityManager = entityManager
         self.eventManager = eventManager
+        self.contacts = []
     }
 
     func update(within time: CGFloat) {
+        let contactComponents = entityManager.components(ofType: ContactComponent.self)
+        var contactsToBeSeparated = contacts
+        contacts = evaluate(contactComponents: contactComponents)
+
         for contact in contacts {
-            handleContact(between: contact.entityIdA, and: contact.entityIdB)
+            contactsToBeSeparated.remove(contact)
         }
-    }
 
-    func insert(contact: TFContact) {
-        guard contact.entityIdA != contact.entityIdB, entityManager?.entity(with: contact.entityIdA) != nil,
-              entityManager?.entity(with: contact.entityIdB) != nil else {
-            return
+        for contact in contactsToBeSeparated {
+            if let entityA = entityManager.entity(with: contact.entityIdA) {
+                handleSeparation(for: entityA)
+            }
+
+            if let entityB = entityManager.entity(with: contact.entityIdB) {
+                handleSeparation(for: entityB)
+            }
         }
-        contacts.insert(contact)
-    }
 
-    func remove(contact: TFContact) {
-        if contacts.remove(contact) != nil {
-            handleSeparation(between: contact.entityIdA, and: contact.entityIdB)
+        for contact in contacts {
+            guard let entityA = entityManager.entity(with: contact.entityIdA),
```

Comment:
so we still using ID?

Reply:
We'll decide on that at a later date, this pr is mainly to resolve the contact system depending on spritekit